src/main/kotlin/org/rust/cargo/project/model/impl/CargoProjectImpl.kt

```diff
@@ -323,8 +331,11 @@ open class CargoProjectsServiceImpl(
      */
     protected fun modifyProjects(
         f: (List<CargoProjectImpl>) -> CompletableFuture<List<CargoProjectImpl>>
-    ): CompletableFuture<List<CargoProjectImpl>> =
-        projects.updateAsync(f)
+    ): CompletableFuture<List<CargoProjectImpl>> {
+        val refreshStatusPublisher = project.messageBus.syncPublisher(CargoProjectsService.CARGO_PROJECTS_REFRESH_TOPIC)
+        refreshStatusPublisher.onRefreshStarted()
```

Comment:
Shouldn't we call the listener from a write action? That is very common in IntelliJ. If not, I think is should be explicitly stated in the `ExternalSystemProjectListener` documentation because it is very unusual.

Reply:
I think this call should be moved into the `.thenApply { projects ->` closure because otherwise 2 serial refresh action invocations will lead to these weird events:
```
RefreshStarted
RefreshStarted
ReloadFinish(CANCEL)
ReloadFinish(SUCCESS)
```

I think it should be fixed even if the default `ExternalSystemProjectListener` implementation is OK with this

Reply:
> Shouldn't we call the listener from a write action? That is very common in IntelliJ. If not, I think is should be explicitly stated in the ExternalSystemProjectListener documentation because it is very unusual.

As I can see, `com.intellij.openapi.externalSystem.autoimport.ProjectSettingsTracker.ProjectListener` is prepared to call from any thread. Also, both reference implementations (common for all external systems like Gradle and Maven one) written by the same developers don't call it neither under write action nor even in EDT so looks like it's ok here

But let's also ask the author of the API - @HackerMadCat

Reply:
I don't know anything about this restrictions on listener API. Also, I cannot find anything in  https://plugins.jetbrains.com/docs/intellij/plugin-listeners.html. 
Every one who expects write, read or EDT context uses special assertions.